File: rules-runtime/src/protocol.ts
/**
 * Wire protocol between the suspect host (Rust) and the TS rule worker.
 *
 * Transport: NDJSON over the worker's stdio — one JSON object per line.
 * Every frame carries a `t` discriminator. Versioned at `hello`; the host
 * kills and restarts the worker on any protocol violation.
 */

export const PROTOCOL_VERSION = 1;

/** Host → worker: handshake with the rule files to load. */
export interface HelloFrame {
  t: "hello";
  protocol: number;
  sdk_version: string;
  workspace_root: string;
  rule_files: string[];
}

/** Worker → host: rules loaded; metadata lets the host pre-select nodes. */
export interface ReadyFrame {
  t: "ready";
  bun: string;
  sdk: string;
  rules: RuleMeta[];
}

export interface RuleMeta {
  /** Unique rule id from `meta.id`. */
  id: string;
  /** File the rule was loaded from. */
  file: string;
  /** Which document kinds the rule evaluates. */
  targets: TargetKind[];
  /** JSONPath `given` selector (point rules). */
  given?: string;
  /** `point` (per-node check) or `walk` (visitors over the document). */
  shape: "point" | "walk";
  /** Visitor keys declared by walk rules (host informational). */
  visitors?: string[];
  /** True when the rule file imports zod (host informational). */
  usesZod?: boolean;
}

export type TargetKind = "spec" | "facts" | "union";

/** Host → worker: run rules over one shipped document. */
export interface EvaluateFrame {
  t: "evaluate";
  run_id: number;
  timeout_ms: number;
  doc_kind: TargetKind;
  doc_uri?: string;
  /** The document, shipped once for the whole run. */
  document: { value: unknown };
  /** Per-point-rule selected pointers, resolved against `document`. */
  selections: Record<string, string[]>;
  /** Per-walk-rule walk entries. */
  walks: Record<string, { root_pointer: string }>;
}

/** Worker → host: batched findings for one run. */
export interface FindingsBatchFrame {
  t: "findings_batch";
  run_id: number;
  findings: FindingFrame[];
}

/** Worker → host: one violation (legacy single-frame path). */
export interface FindingFrame {
  t: "finding";
  run_id: number;
  rule_id: string;
  pointer: string;
  message: string;
  severity?: "error" | "warning" | "info" | "hint";
  /** Optional structured quick-fix intent (host executes it). */
  fix?: FixIntent;
}

export interface FixIntent {
  kind: "insert-doc" | "document-in-spec" | "note";
  template?: string;
  message?: string;
}

/** Worker → host: a run finished. */
export interface DoneFrame {
  t: "done";
  run_id: number;
  ms: number;
  findings: number;
}

/** Worker → host: a rule threw; host disables it for the session. */
export interface RuleErrorFrame {
  t: "rule_error";
  run_id: number;
  rule_id: string;
  message: string;
}

/** Worker → host: fatal worker problem (protocol/bun level). */
export interface FatalFrame {
  t: "fatal";
  message: string;
}

/** Host → worker: re-import changed rule files. */
export interface ReloadFrame {
  t: "reload";
  files: string[];
}

/** Liveness probes. */
export interface PingFrame {
  t: "ping";
}
export interface PongFrame {
  t: "pong";
}

export type HostFrame = HelloFrame | EvaluateFrame | ReloadFrame | PingFrame;
export type WorkerFrame =
  | ReadyFrame
  | FindingsBatchFrame
  | FindingFrame
  | DoneFrame
  | RuleErrorFrame
  | FatalFrame
  | PongFrame;

File: crates/suspect-rules/src/protocol.rs
//! Wire frames shared with `rules-runtime/src/protocol.ts`. One JSON object
//! per line; every frame carries a `t` discriminator (applied via serde
//! enum tags).

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Protocol version; bumped on any breaking frame change.
pub const PROTOCOL_VERSION: u32 = 1;

/// Which document kind a rule (or run) evaluates.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TargetKind {
    /// The OpenAPI document.
    Spec,
    /// Code-extraction facts (fold: `suspect-code`).
    Facts,
    /// Spec ∪ facts merged graph.
    Union,
}

/// Worker → host: per-rule metadata; the host compiles `given` natively.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RuleMeta {
    /// Unique rule id from `meta.id`.
    pub id: String,
    /// File the rule was loaded from.
    pub file: String,
    /// Which document kinds the rule evaluates.
    pub targets: Vec<TargetKind>,
    /// JSONPath `given` selector (point rules).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub given: Option<String>,
    /// Point (per-node `check`) or walk (visitor traversal) shape.
    pub shape: Shape,
    /// Visitor keys declared by walk rules (host informational).
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub visitors: Vec<String>,
    /// True when the rule file imports zod (host informational).
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub uses_zod: bool,
}

/// Point (per-node `check`) or walk (visitor traversal) rule shape.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Shape {
    /// `given` + `check(node, ctx)` per selected node.
    Point,
    /// Visitors traversing the document with shared state.
    Walk,
}

/// Document root shipped once per evaluate run; point rules receive
/// pointer lists and resolve against it worker-side. The value is raw JSON
/// text (`RawValue`) so the host never re-serializes it inside the frame.
#[derive(Debug, Clone)]
pub struct RunDocument {
    /// Whole document as pre-serialized JSON (spliced verbatim into the
    /// frame — never re-serialized host-side).
    pub value: Box<serde_json::value::RawValue>,
}

impl Serialize for RunDocument {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        // Shape must stay `{"value": <raw doc>}` — the worker reads
        // `frame.document.value`. RawValue splices the JSON verbatim.
        use serde::ser::SerializeMap;
        let mut map = serializer.serialize_map(Some(1))?;
        map.serialize_entry("value", &self.value)?;
        map.end()
    }
}

/// Walk-rule entry: the shared document plus the walk start.
#[derive(Debug, Clone, Serialize)]
pub struct WalkEntry {
    /// Pointer the walk starts from (usually empty).
    pub root_pointer: String,
}

/// Worker → host: one violation.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FindingBody {
    /// Run the finding belongs to.
    pub run_id: u64,
    /// Producing rule.
    pub rule_id: String,
    /// RFC 6901 pointer of the offending node.
    pub pointer: String,
    /// Human message.
    pub message: String,
    /// Severity override (rule default when absent).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub severity: Option<String>,
    /// Structured fix intent (pass-through; host executes).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fix: Option<serde_json::Value>,
}

/// Worker → host: batched findings for one run (the worker buffers a
/// run's findings and emits them in one frame to avoid per-frame
/// round-trip costs at finding volume).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FindingsBatchBody {
    /// Run the batch belongs to.
    pub run_id: u64,
    /// All findings for the run, emission order.
    pub findings: Vec<FindingBody>,
}

/// Worker → host: a rule threw during a run.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RuleErrorBody {
    /// Run the error occurred in.
    pub run_id: u64,
    /// Rule that threw.
    pub rule_id: String,
    /// Error message.
    pub message: String,
}

/// Worker → host: run completed.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct DoneBody {
    /// Completed run id.
    pub run_id: u64,
    /// Worker-measured duration in milliseconds.
    pub ms: f64,
    /// Findings emitted during the run.
    pub findings: u32,
}

/// Payload of [`WorkerFrame::Ready`].
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReadyPayload {
    /// Bun version string.
    pub bun: String,
    /// SDK version the worker compiled against.
    pub sdk: String,
    /// Loaded rules in declaration order.
    pub rules: Vec<RuleMeta>,
}

/// Payload of [`WorkerFrame::Fatal`].
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FatalBody {
    /// What went wrong, worker-side.
    pub message: String,
}

/// Worker → host: any frame the worker sends.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "t", rename_all = "snake_case")]
pub enum WorkerFrame {
    /// Rules loaded after hello/reload.
    Ready(ReadyPayload),
    /// One violation (legacy single-frame path).
    Finding(FindingBody),
    /// Batched findings for a run.
    FindingsBatch(FindingsBatchBody),
    /// Run finished.
    Done(DoneBody),
    /// A rule threw.
    RuleError(RuleErrorBody),
    /// Fatal worker problem.
    Fatal(FatalBody),
    /// Liveness reply.
    Pong,
}

/// Host → worker frames, tagged the same way.
#[derive(Debug, Clone, Serialize)]
#[serde(tag = "t", rename_all = "snake_case")]
pub enum HostFrame {
    /// Handshake: protocol version, workspace, rule files to load.
    Hello {
        /// Wire protocol version.
        protocol: u32,
        /// Host-side SDK version.
        sdk_version: String,
        /// Workspace root (worker cwd, read jail).
        workspace_root: String,
        /// Absolute rule file paths to import.
        rule_files: Vec<String>,
    },
    /// Evaluate: one document payload + per-rule pointer lists.
    Evaluate {
        /// Monotonic run id.
        run_id: u64,
        /// Deadline the host enforces.
        timeout_ms: u64,
        /// Document kind under evaluation.
        doc_kind: TargetKind,
        /// Document uri when known.
        #[serde(skip_serializing_if = "Option::is_none")]
        doc_uri: Option<String>,
        /// The document, shipped once for the whole run.
        document: RunDocument,
        /// Per-point-rule selected pointers.
        selections: BTreeMap<String, Vec<String>>,
        /// Per-walk-rule walk entries.
        walks: BTreeMap<String, WalkEntry>,
    },
    /// Re-import changed rule files.
    Reload {
        /// Files that changed.
        files: Vec<String>,
    },
    /// Liveness probe.
    Ping,
}
